File: packages/ts/src/application/ports/gate-policy.ts
import type { RunState } from "../../domain/run";
import type { Tool, ToolCall } from "../../domain/tool";

export interface GateDecisionInput {
  readonly tool: Tool;
  readonly toolCall: ToolCall;
  readonly state: RunState;
}

/**
 * Decides whether a specific tool call needs human approval. Kept out of tool
 * definitions and out of the prompt, so the rule is deterministic and testable
 * (see ADR-0003).
 */
export interface GatePolicy {
  requiresApproval(input: GateDecisionInput): boolean;
}

File: packages/py/src/airlock/application/ports/gate_policy.py
"""GatePolicy port: decides whether a specific tool call needs human approval.

Kept out of tool definitions and the prompt, so the rule is deterministic and
testable (ADR-0003)."""

from dataclasses import dataclass
from typing import Protocol

from airlock.domain.run import RunState
from airlock.domain.tool import Tool, ToolCall


@dataclass(frozen=True)
class GateDecisionInput:
    tool: Tool
    tool_call: ToolCall
    state: RunState


class GatePolicy(Protocol):
    def requires_approval(self, decision_input: GateDecisionInput) -> bool: ...
